### PR TITLE
Prevent associative arrays and maps of non-nilable classes

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR makes the initialization of associative arrays of non-nilable classes a compile-time error. It does so by adding a postinit for the `DefaultAssociativeArr` class which issues a `compilerError` if the element type of the array is a non-nilable class.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [ ] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Quite a few errors have popped up from code that leans on associative arrays of non-nilable classes.

**Comment by dlongnecke-cray** [comment]
It turns out that there is quite a bit of code in distributions that relies on being able to make an associative array of locales. This in turn triggers a lot of test errors when we blanket ban associative arrays of non-nilable classes. Do we want to make an exception if the class type is a `locale`? Or should I go through the effort of modifying distribution code to use the new nilability features and change the type to `locale?`.

**Comment by dlongnecke-cray** [comment]
I've decided to go with the route of making an exception for the `locale` type in `postinit`, as Michael has informed me it is also an exception in the lifetime checker.

**Comment by dlongnecke-cray** [comment]
I've encountered another problem which I didn't think about initially, which is what workaround to use here for the `map` type.

The `map` type uses an associative array for its implementation. However unlike an associative array the `map` is safe to use with non-nilable classes because key/values can be initialized together at the same time (there is no need for a default value).

I figure we have two options here:

a) Rewrite `map` to not lean on associative arrays (not very feasible)
b) Provide a workaround so that map can use associative arrays without emitting the compiler error

Of the two I think `b)` is the correct way to go, but I'm not sure of how to approach it.

**Comment by mppf** [comment]
@dlongnecke-cray - it would be good to make an issue asking if/how map should support non-nilable class elements (if you haven't already).

**Comment by dlongnecke-cray** [comment]
I've made an issue about supporting `map` with a value type of non-nilable class here: https://github.com/chapel-lang/chapel/issues/14861

**Comment by dlongnecke-cray** [comment]
Almost done with this effort, just a few more tests to go. One difficult test with mason, rest are just to do with SSCA.

**Comment by dlongnecke-cray** [comment]
Only two failing tests left - mason dependencies are not being calculated correctly at runtime. I think this effort is ready for review, though.

**Comment by dlongnecke-cray** [comment]
Bug is fixed! A call to `.set` with a list of non-nilable Toml was being dropped on the floor. Extra overloads fixed the problem.

**Comment by dlongnecke-cray** [comment]
Ok, I've done final code cleanups and am running gasnet testing now.

File: test/arrays/errors/nilStorageNonNilAssocArr.chpl [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+type UnmanagedC = unmanaged C;
+
+config type TestClass = OwnedC;
+
+proc main() {
+  var d: domain(string);
+  var a: [d] TestClass;
+
+  // Some sort of error should have been emitted by this point.
+  a += "key";
```
Comment (daviditen) [inline test/arrays/errors/nilStorageNonNilAssocArr.chpl]: I think you meant `d += "key";` here.

File: test/studies/amr/advection/level/LevelBC_AdvectionCTU.chpl [hunk 2]
```diff
@@ -13,9 +13,9 @@ class ZeroInflowBC: LevelBC {
 
     for grid in level.grids {
 
-      for ghost_domain in level.boundary(grid) {
+      for ghost_domain in level.boundary(grid)! {
```
Comment (daviditen) [inline test/studies/amr/advection/level/LevelBC_AdvectionCTU.chpl]: What does this `!` on the iterator expression do?

File: test/studies/amr/lib/cfboundary/RefiningTransfer.chpl [hunk 3]
```diff
@@ -294,7 +294,7 @@ class GridCFGhostSolution {
       //---- This is a zipper iteration over two objects ----
 
       for ( old_array,            current_array            ) 
-      in  zip( old_data(c_neighbor), current_data(c_neighbor) )
+      in  zip( old_data(c_neighbor)!, current_data(c_neighbor)! )
```
Comment (daviditen) [inline test/studies/amr/lib/cfboundary/RefiningTransfer.chpl]: More `!` on iterator expressions here. Not sure if correct or not. There are others in this file also.

File: test/studies/amr/advection/level/LevelBC_AdvectionCTU.chpl (hunk 2 above)
Comment (dlongnecke-cray) [reply test/studies/amr/advection/level/LevelBC_AdvectionCTU.chpl]: It looks like it promotes a nilable element to a non-nilable, but I'm not sure if that's intentional or not (?).

Comment (dlongnecke-cray) [reply test/studies/amr/advection/level/LevelBC_AdvectionCTU.chpl]: It does work, though.

File: test/arrays/errors/nilStorageNonNilAssocArr.chpl (hunk 1 above)
Comment (dlongnecke-cray) [reply test/arrays/errors/nilStorageNonNilAssocArr.chpl]: Good catch, thanks!

**Pull request merged**